### PR TITLE
Test to verify TCS exhaustion to address missing thread binding coverage 

### PR DESCRIPTION
**Pull request by anitagov**
New thread binding test coverage to verify TCS exhaustion i.e. enter on N threads when there are M TCSes where N > M 
oe_call should return OE_OUT_OF_THREADS when M enclaves are in use.
Trick is to keep the M enclaves busy until we get TCS exhaustion and run out of all the threads.
This was mentioned in Issue #24

### CODE REVIEW

**Comment by anitagov** [comment]
My sand-box run passed.

**Comment by anitagov** [comment]
https://oe-jenkins.eastus.cloudapp.azure.com/blue/organizations/jenkins/OpenEnclave-sandbox/detail/OpenEnclave-sandbox/256/pipeline/ => My sand-box run which just passed.

File: tests/thread/host/host.cpp [hunk 1]
```diff
@@ -4,18 +4,34 @@
...
 
 static TestMutexArgs _args;
+static TestTCSArgs _tcsargs;
+static std::atomic_flag _host_tcs_lock = ATOMIC_FLAG_INIT;
 
 const size_t NUM_THREADS = 8;
 
+static inline void _acquire_lock(std::atomic_flag* lock)
+{
+    while (lock->test_and_set(std::memory_order_acquire))
```
Comment (anakrish) [inline tests/thread/host/host.cpp]: Nice use of acquire release.

File: tests/thread/oethread_enc/enc.cpp [hunk 2]
```diff
@@ -245,6 +245,20 @@ OE_ECALL void LockAndUnlockMutexes(void* arg)
...
+OE_ECALL void TestTCSExhaustion(void* args_)
+{
+    TestTCSArgs* volatile args = (TestTCSArgs*)args_;
+    static oe_spinlock_t _tcs_lock = OE_SPINLOCK_INITIALIZER;
+
+    // Increment the number of threads only on getting the _tcs_lock
+    oe_spin_lock(&_tcs_lock);
+    args->num_threads++;
+    oe_spin_unlock(&_tcs_lock);
+    while (args->num_threads + args->num_out_threads < args->tcs_count)
```
Comment (anakrish) [inline tests/thread/oethread_enc/enc.cpp]: Can you describe the logic of this loop here?
It is not obvious what it is doing.

File: tests/thread/host/host.cpp [hunk 3]
```diff
@@ -232,6 +250,75 @@ void TestThreadLockingPatterns(oe_enclave_t* enclave)
...
+
+    for (size_t i = 0; i < test_tcs_count; i++)
+        threads[i].join();
+
+    printf(
+        "TestTCSExhaustion(): tcs_count=%d; num_threads=%d; "
+        "num_out_threads=%d\n",
+        (int)test_tcs_count,
+        (int)_tcsargs.num_threads,
+        (int)(int)_tcsargs.num_out_threads);
```
Comment (anakrish) [inline tests/thread/host/host.cpp]: Remove double cast

File: tests/thread/oethread_enc/enc.cpp (hunk 2 above)
Comment (anakrish) [reply tests/thread/oethread_enc/enc.cpp]: Maybe rename the num_threads, num_out_threads?
Initially I thought num_threads was total number of threads.

File: tests/thread/host/host.cpp (hunk 1 above)
Comment (anitagov) [reply tests/thread/host/host.cpp]: Thanks

File: tests/thread/oethread_enc/enc.cpp (hunk 2 above)
Comment (anitagov) [reply tests/thread/oethread_enc/enc.cpp]: ok, will incorporate

File: tests/thread/host/host.cpp (hunk 3 above)
Comment (anitagov) [reply tests/thread/host/host.cpp]: Fixed.

File: tests/thread/oethread_enc/enc.cpp (hunk 2 above)
Comment (anitagov) [reply tests/thread/oethread_enc/enc.cpp]: Changed the names of the variables to reflect the tcs bindings being used and requested. Also added comment to describe the logic of the loop. High level description is also in host.cpp before the routine.

**Comment by anitagov** [comment]
bors try

**Comment by anitagov** [comment]
bors try

**Comment by anitagov** [comment]
bors try

**Comment by anitagov** [comment]
Passed in sandbox mode. Retrying bors try

**Comment by anitagov** [comment]
bors try

**Comment by anitagov** [comment]
bors try

**Comment by anitagov** [comment]
bors try

**Comment by johnkord** [comment]
Oops disregard my comment

**Pull request merged**